metaseq/cli/inference.py

```diff
@@ -0,0 +1,764 @@
+
+        if not args.use_cpu:
+            batch = utils.move_to_cuda(batch)
+
+        translate_start_time = time.time()
+        output_generation = generator_interface.task.inference_step(generator, generator_interface.models, batch)
+        translate_time = time.time() - translate_start_time
+        total_generation_time += translate_time
+
+        all_tokens = output_generation["tokens"].cpu()
+        all_scores = output_generation["scores"].cpu()
+        if args.top_k_logprobs > 0:
+            all_distributions = output_generation["distributions"].cpu(
+            )  # (bsz, beam_size, seq_len + prompt_size, self.vocab_size)
+        else:
+            all_distributions = None
+
+        if args.top_k_logits > 0:
+            all_logits = output_generation["logits"].cpu()  # (bsz, beam_size, seq_len + prompt_size, self.vocab_size)
+        else:
+            all_logits = None
+
+        batch_itemgetter = operator.itemgetter(*batch["id"].tolist())
+        batch_target_text = list(batch_itemgetter(tgt_text)) if tgt_text[0] is not None else []
+        batch_input_items = list(batch_itemgetter(input_items)) if args.copy_input_to_output else []
+
+        # Async write to file
+        async_processor.submit(
+            write_inference_results_to_file,
+            output_file=os.path.join(
+                str(cfg.common_eval.results_path),
+                f"worker_prediction_results_rank{distributed_utils.get_data_parallel_rank()}.jsonl"
+            ),
+            generator_interface=generator_interface,
+            generator=generator,
+            indexs=batch["id"],
+            src_lengths=src_lengths,
+            all_tokens=all_tokens,
+            all_scores=all_scores,
+            all_distributions=all_distributions,
+            all_logits=all_logits,
+            best_n=args.best_n,
+            num_logprobs=args.top_k_logprobs,
+            num_logits=args.top_k_logits,
+            target_text=batch_target_text,
+            input_items=batch_input_items,
+            copy_input_to_output=args.copy_input_to_output,
+            input_keys_to_copy=args.input_keys_to_copy,
+            output_tokens_and_offsets=args.output_tokens_and_offsets,
+            echo_prompt=args.echo_prompt,
+            progress_bar=progress_bar,
+            translate_time=translate_time,
+        )
+
+    # All processes wait for the async writes to finish at the end
+    async_processor.shutdown()
+    distributed_utils.global_barrier()
+
+    if distributed_utils.get_global_rank() == 0:
+        progress_bar.close()
+
+        if args.merge_preds_on_all_ranks:
+            unified_output_prediction_file = os.path.join(str(cfg.common_eval.results_path), "all_prediction_results.jsonl")
+            # TODO this means that files are not sorted! In previous
+            # implementation the predictions were sorted by `instance_idx`.
+            # Confirm if this is really necessary (it isn't as far as I've seen)
```

Comment:
I we can remove this comment since it's no longer relevant. If I'm not mistaken, it's a leftover from one of my PRs and I never intended to merge it 😅

```suggestion
```

Reply:
Ok, removed